compiler/sem/semstmts.nim

```diff
@@ -2944,24 +2993,151 @@ proc semConverterDef(c: PContext, n: PNode): PNode =
-    localReport(c.config, n, reportSym(
-      rsemImplementationExpected, s))
+
+  result = shallowCopy(n)
+  # copy over the original nodes at the start already. This makes returning
+  # early in case of an error easier
+  for i in 0..<n.len:
+    result[i] = n[i]
+
+  var
+    s = n[namePos].sym
+    hasError = false
+
+  s.ast = result
+  s.options = c.config.options # captue the current options
+
+  let declarationScope = c.currentScope
+  # the scope and owner need to apply to the parameters too, so push and open
+  # them already
+  pushOwner(c, s)
+  openScope(c)
+
+  s.typ = semRoutineParams(c, result, n[paramsPos], n[genericParamsPos], skMacro)
+  if s.typ == nil:
+    s.typ = newProcType(c, n.info)
+
+  # ----- compute symbol flags -------
+
+  # are all parameters untyped?
+  incl(s.flags, sfAllUntyped) # assume they are
+  for i in 1..<s.typ.n.len:
+    if s.typ.n[i].sym.typ.kind != tyUntyped:
+      excl(s.flags, sfAllUntyped) # they weren't
+      break
+
+  # macros are always compile-time-only routines
+  incl(s.flags, sfCompileTime)
+
+  # now analyse the pattern slot. This has to happen *after* the parameters
+  # were analysed, as the pattern might reference them
+  result[patternPos] = semPattern(c, n[patternPos])
+
+  if result[patternPos].isError:
+    hasError = true
+
+  # register the macro in the symbol table and owning module's interface (if
+  # the owner is a module)
+  if sfGenSym notin s.flags:
+    addInterfaceOverloadableSymAt(c, declarationScope, s)
+
+  # apply the normal pragmas:
+  result[pragmasPos] = pragmaCallable2(c, s, n[pragmasPos], macroPragmas)
+  if not result[pragmasPos].isError:
+    s = implicitPragmas(c, s, n.info, macroPragmas)
+
+  if s.isError: # did application fail?
+    result[namePos] = s.ast
+    hasError = true
+  else:
+    result[namePos] = checkSpecialOperators(c, result[namePos])
+    hasError = hasError or result[namePos].kind == nkError
+
+  if hasError:
+    # don't analyse the body if the header has errors:
+    closeScope(c)
+    popOwner(c)
+    return wrapError(c.config, result)
+
+  if result[pragmasPos].kind != nkEmpty:
+    setEffectsForProcType(c.graph, s.typ, result[pragmasPos], s)
+  s.typ.flags.incl tfEffectSystemWorkaround
+
+  # analyse the body:
+  if n[bodyPos].kind != nkEmpty and sfError notin s.flags:
+    pushProcCon(c, s)
+    maybeAddResult(c, s, result)
+    result[bodyPos] = hloBody(c, semProcBody(c, n[bodyPos]))
+    trackProc(c, s, result[bodyPos])
+    popProcCon(c)
+
+    hasError = result[bodyPos].isError
+  elif n[bodyPos].kind == nkEmpty:
+    result[bodyPos] =
+      newError(c.config, n[bodyPos],
+               PAstDiag(kind: adSemImplementationExpected,
+                        routineSym: s,
+                        routineDefStartPos: n.info))
+
+    hasError = true
+
+  closeScope(c)
+  popOwner(c)
+
+  if hasError:
+    # shortcut pattern registration
+    # QUESTION: transition to error symbol?
+    return c.config.wrapError(result)
```

Comment:
I think this might be required because of the way some of the lookups happen during `sigmatch`, but it's more a defensive measure and I didn't get to ground on why that was the case. I say leave it for now -- worst case it'll surface some bugs.